java/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/EventData.java

```diff
@@ -195,53 +179,116 @@ public EventData(ByteBuffer buffer)
 
 	/**
 	 * Get Actual Payload/Data wrapped by EventData.
-	 * This is intended to be used after receiving EventData using @@PartitionReceiver.
-	 * @return returns the byte[] of the actual data 
+	 * This is the underlying array and should be used in conjunction with {@link #getBodyOffset()} & {@link #getBodyLength()}.
+ 	 * @return returns the byte[] of the actual data
 	 */
 	public byte[] getBody()
 	{
-		// TODO: enforce on-send constructor type 2
 		return this.bodyData == null ? null : this.bodyData.getArray();
 	}
+	
+	/**
+	 * Get the offset of the current Payload/Data in the byte array returned by {@link #getPayloadArray()}.
+	 * @return returns the byte[] of the actual data
+	 * @see #getBodyLength()
+	 * @see #getBody()
+	 */
+	public int getBodyOffset()
+	{
+		return this.bodyData == null ? 0 : this.bodyData.getArrayOffset();
+	}
+	
+	/**
+	 * Get the length of the Actual Payload/Data in the byte array returned by {@link #getPayloadArray()}.
+	 * @return returns the byte[] of the actual data
+	 * @see #getBody()
+	 * @see #getBodyOffset()
+	 */
+	public int getBodyLength()
+	{
+		return this.bodyData == null ? 0 : this.bodyData.getLength();
+	}
 
 	/**
 	 * Application property bag
 	 * @return returns Application properties
 	 */
 	public Map<String, String> getProperties()
 	{
-		return this.properties;
-	}
+		if (this.properties == null)
+		{
+			this.properties = new HashMap<String, String>();
+		}
 
-	public void setProperties(Map<String, String> applicationProperties)
-	{
-		this.properties = applicationProperties;
+		return this.properties;
 	}
-
+	
 	/**
 	 * SystemProperties that are populated by EventHubService.
 	 * <p>As these are populated by Service, they are only present on a Received EventData.
-	 * @return an encapsulation of all SystemProperties appended by EventHubs service into EventData
+	 * <p>Usage:<p>
+	 * <code>
+	 * final String offset = eventData.getSystemProperties().getOffset();
+	 * </code>
+	 * @return an encapsulation of all SystemProperties appended by EventHubs service into EventData.
+	 * <code>null</code> if the {@link #EventData()} is not received and is created by the public constructors.
+	 * @see SystemProperties#getOffset
+	 * @see SystemProperties#getSequenceNumber
+	 * @see SystemProperties#getPartitionKey
+	 * @see SystemProperties#getEnqueuedTime
 	 */
 	public SystemProperties getSystemProperties()
 	{
-		if (this.isReceivedEvent && this.systemProperties == null)
-		{
-			this.systemProperties = new SystemProperties(this);
-		}
-
 		return this.systemProperties;
```

Comment:
>return this.systemProperties; [](start = 2, length = 29)

how are we enforcing this to be only showing up in received EventData? I thought that was what "isReceivedEvent" is for but I don't see it being used here.